File: backend/internal/models/experience.go
// ============================================================
// FILE: internal/models/experience.go
// WHAT IT IS:     Work experience database model
// WHY IT EXISTS:  Stores each job entry shown in the Experience
//                 Timeline section (SperkTech, SoftVence, DIU)
// ENDPOINTS:      Used by experience_handler.go
// DEPENDS ON:     gorm.io/gorm
// IF REMOVED:     Experience Timeline section has no data
// LAST UPDATED:   2026-05-07 — initial creation
// ============================================================

package models

import (
	"time"

	"gorm.io/gorm"
)

// Experience represents one job/internship on the timeline.
// Maps to the "experiences" table in PostgreSQL.
type Experience struct {
	// Primary key
	ID uint `gorm:"primaryKey;autoIncrement" json:"id"`

	// Company is the employer name (e.g. "SperkTech")
	Company string `gorm:"type:varchar(200);not null" json:"company"`

	// Role is the job title (e.g. "Flutter Developer")
	Role string `gorm:"type:varchar(200);not null" json:"role"`

	// StartDate is when the job began (stored as string "Jan 2022" for display)
	StartDate string `gorm:"type:varchar(50);not null" json:"start_date"`

	// EndDate is when the job ended — empty string means "Present"
	EndDate string `gorm:"type:varchar(50)" json:"end_date"`

	// IsCurrent marks the most recent job so the UI shows "Present" badge
	IsCurrent bool `gorm:"default:false" json:"is_current"`

	// Description is the main paragraph explaining the role
	Description string `gorm:"type:text" json:"description"`

	// Achievements stores the bullet points of specific accomplishments
	// Stored as JSON array: ["Built X", "Improved Y by Z%"]
	Achievements StringArray `gorm:"type:text" json:"achievements"`

	// TechUsed stores the technology tags shown on the timeline card
	// Example: ["Flutter", "Dart", "Firebase", "GetX"]
	TechUsed StringArray `gorm:"type:text" json:"tech_used"`

	// CompanyLogo is the path to the company's logo image
	CompanyLogo string `gorm:"type:varchar(500)" json:"company_logo"`

	// Location is where the job was (e.g. "Dhaka, Bangladesh")
	Location string `gorm:"type:varchar(200)" json:"location"`

	// Type is the employment type: "full-time" | "part-time" | "internship" | "freelance"
	Type string `gorm:"type:varchar(50);default:'full-time'" json:"type"`

	// SortOrder controls which experience shows first (lower = more recent, shown first)
	SortOrder int `gorm:"default:0" json:"sort_order"`

	// ─── Per-section visibility flags ──────────────────────────
	// Each flag controls whether the corresponding section renders on the
	// public timeline. Stored as *bool (not bool) so GORM's Updates() can
	// persist a `false` toggle — Updates skips zero-value fields, and
	// `false` is the zero value of a plain bool, so plain-bool toggles get
	// silently dropped on save. Null/missing on legacy rows is treated as
	// "visible" by the frontend, matching the column default.
	ShowLogo         *bool `gorm:"default:true" json:"show_logo"`
	ShowType         *bool `gorm:"default:true" json:"show_type"`
	ShowLocation     *bool `gorm:"default:true" json:"show_location"`
	ShowDescription  *bool `gorm:"default:true" json:"show_description"`
	ShowAchievements *bool `gorm:"default:true" json:"show_achievements"`
	ShowTechUsed     *bool `gorm:"default:true" json:"show_tech_used"`

	// GORM timestamp fields
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
}

File: frontend/assets/js/pages/portfolio/experience.js
/* ============================================================
 * FILE: pages/portfolio/experience.js
 * WHAT IT IS:     Experience Timeline section loader
 * WHERE USED:     main.js → called on DOMContentLoaded
 * LAST UPDATED:   2026-05-07 — initial creation
 * ============================================================ */

const ExperienceSection = (() => {

  async function init() {
    const container = document.getElementById('experience-timeline');
    if (!container) return;

    try {
      const res = await API.get('/experience');
      const experiences = res.data || [];

      if (!experiences.length) {
        renderFallback(container);
        return;
      }

      container.innerHTML = experiences.map(exp => buildTimelineItem(exp)).join('');
    } catch {
      renderFallback(container);
    }
  }

  // Per-section visibility helper. Backend stores show_* as *bool, so the
  // value on the wire can be true / false / null. null (and undefined for
  // legacy rows saved before these flags existed) is treated as visible —
  // matching the DB column default. Only an explicit `false` hides.
  const isVisible = (val) => val !== false;

  function buildTimelineItem(exp) {
    const period = exp.end_date ? `${exp.start_date} — ${exp.end_date}` : exp.start_date;

    const logo = (isVisible(exp.show_logo) && exp.company_logo)
      ? `<img class="timeline-logo" src="${exp.company_logo}" alt="${exp.company} logo"
           onerror="this.style.display='none'">`
      : '';

    const locationSuffix = (isVisible(exp.show_location) && exp.location)
      ? ` · ${exp.location}` : '';

    const typeBadge = isVisible(exp.show_type)
      ? `<span class="timeline-type ${exp.type || 'full-time'}">${exp.type || 'Full-time'}</span>`
      : '';

    const description = (isVisible(exp.show_description) && exp.description)
      ? `<p style="font-size:var(--fs-sm);color:var(--color-text-muted);line-height:1.7;margin-bottom:1rem">${exp.description}</p>`
      : '';

    const achievements = isVisible(exp.show_achievements)
      ? Utils.parseArray(exp.achievements).map(a => `<li>${a}</li>`).join('')
      : '';

    const techTags = isVisible(exp.show_tech_used)
      ? Utils.parseArray(exp.tech_used).map(t => `<span class="timeline-tech-tag">${t}</span>`).join('')
      : '';

    return `
      <div class="timeline-item scroll-animation" data-animation="fade_from_bottom">
        <div class="timeline-header">
          ${logo}
          <div class="timeline-company-info">
            <h3 class="timeline-role">${exp.role}</h3>
            <div class="timeline-company">${exp.company}${locationSuffix}</div>
          </div>
          <div class="timeline-meta">
            <span class="timeline-period">${period}</span>
            ${exp.is_current ? `<span class="timeline-current">Present</span>` : ''}
            ${typeBadge}
          </div>
        </div>
        ${description}
        ${achievements ? `<ul class="timeline-achievements">${achievements}</ul>` : ''}
        ${techTags ? `<div class="timeline-tech">${techTags}</div>` : ''}
      </div>
    `;
  }

  function renderFallback(container) {
    container.innerHTML = `<p style="color:var(--color-text-muted)">Experience data will appear here once the backend is connected.</p>`;
  }

  return { init };
})();
